components/ui/pathogen-map/pathogen-map-popup.tsx

```diff
@@ -0,0 +1,102 @@
+> = (input: TPathogenDataPointProperties) => React.ReactNode;
+
+type VisiblePopupInfo<
+  TPathogenDataPointProperties extends PathogenDataPointPropertiesBase
+> = { visible: true; properties: TPathogenDataPointProperties };
+type HiddenPopupInfo<
+  TPathogenDataPointProperties extends PathogenDataPointPropertiesBase
+> = { visible: false; properties: null };
+
+export type PopupInfo<
+  TPathogenDataPointProperties extends PathogenDataPointPropertiesBase
+> =
+  | VisiblePopupInfo<TPathogenDataPointProperties>
+  | HiddenPopupInfo<TPathogenDataPointProperties>;
+
+interface PathogenMapPopupProps<
+  TPathogenDataPointProperties extends PathogenDataPointPropertiesBase
+> {
+  mapId: string;
+  popUpInfo: PopupInfo<TPathogenDataPointProperties>;
+  generatePopupContent: PopupContentGenerator<TPathogenDataPointProperties>;
+}
+
+export function PathogenMapPopup<
+  TPathogenDataPointProperties extends PathogenDataPointPropertiesBase
+>({
+  mapId,
+  popUpInfo,
+  generatePopupContent,
+}: PathogenMapPopupProps<TPathogenDataPointProperties>) {
+  const allMaps = useMap();
+  const map = allMaps[mapId];
+
+  const getMapboxLatitudeOffset = (map: MapRef) => {
+    // Map seems to zoom in in powers of 2, so reducing offset by powers of 2 keeps the modal apprximately
+    // in the same center everytime
+    // offset needs to reduce exponentially with zoom -- higher zoom x smaller offset
+    let mapZoom = map.getZoom();
+
+    return 80 / Math.pow(2, mapZoom);
+  };
+
+  if (popUpInfo.visible === false) {
+    return null;
+  }
+
+  // Not a typo, there is some kind of underlying problem here worth investigating that
+  // makes switching the latitude and longitude a required step to produce the correct results.
+  const latitude = popUpInfo.properties.longitude ?? 0;
+  const longitude = popUpInfo.properties.latitude ?? 0;
+
+  const transformedProperties = Object.fromEntries(
+    Object.entries(popUpInfo.properties).map(([key, value]) => {
+      // Arrays in popUpInfo.properties that reach this point have been converted from
+      // ["ABC", "DEF"] to '["ABC", "DEF"]'
+      // In order to proceed further, they need to be converted back to arrays.
+      if (/^\[("[^\[\]\"]*",?)+\]$/.test(value)) {
+        return [
+          key,
+          value
+            .slice(1, -1)
+            .split(",")
+            .map((innerString: string) => innerString.slice(1, -1)),
+        ];
+      }
+
+      return [key, value];
+    })
+  ) as VisiblePopupInfo<TPathogenDataPointProperties>["properties"];
+
+  return (
+    <Popup
+      // Important, without this key, clicking from one popup to another does not trigger the onOpen since react thinks the component data
+      // has changed instead of a new pop up being opened. If you choose to change this, please ensure that in your solution the map flys to
+      // the correct location when clicking on one data point while having the popup already open.
+      key={transformedProperties.id}
```

Comment:
If you're wondering why I didn't move the `map.flyTo` to `onMouseDown` if this is the case, for some reason it doesn't work if you put it in the `onMouseDown`. I tried for ages to make it work in the `onMouseDown` and I swear it's some kind of janky mapbox interaction because I'm convinced I was calling everything correctly. Honestly this looks cleaner anyways imo just leaving a comment in case you were curious.